sw/device/lib/dif/dif_clkmgr.h

```diff
@@ -0,0 +1,200 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_CLKMGR_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_CLKMGR_H_
+
+/**
+ * @file
+ * @brief <a href="/hw/ip/clkmgr/doc/">Clock manager</a> Device Interface
+ * Functions
+ */
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_warn_unused_result.h"
+
+// Header Extern Guard (so header can be used from C and C++)
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * A toggle state: enabled, or disabled.
+ *
+ * This enum may be used instead of a `bool` when describing an enabled/disabled
+ * state.
+ */
+typedef enum dif_clkmgr_toggle {
+  /*
+   * The "enabled" state.
+   */
+  kDifClkmgrToggleEnabled,
+  /**
+   * The "disabled" state.
+   */
+  kDifClkmgrToggleDisabled,
+} dif_clkmgr_toggle_t;
+
+/**
+ * Hardware instantiation parameters for clock manager.
+ *
+ * This struct describes information about the underlying hardware that is
+ * not determined until the hardware design is used as part of a top-level
+ * design.
+ */
+typedef struct dif_clkmgr_params {
+  /**
+   * The base address for the clock manager hardware registers.
+   */
+  mmio_region_t base_addr;
+} dif_clkmgr_params_t;
+
+/**
+ * A handle to clock manager.
+ *
+ * This type should be treated as opaque by users.
+ */
+typedef struct dif_clkmgr {
+  dif_clkmgr_params_t params;
+} dif_clkmgr_t;
+
+/**
+ * The result of a clock manager operation.
+ */
+typedef enum dif_clkmgr_result {
+  /**
+   * Indicates that the operation succeeded.
+   */
+  kDifClkmgrOk = 0,
+  /**
+   * Indicates some unspecified failure.
+   */
+  kDifClkmgrError = 1,
+  /**
+   * Indicates that some parameter passed into a function failed a
+   * precondition.
+   *
+   * When this value is returned, no hardware operations occured.
+   */
+  kDifClkmgrBadArg = 2,
+} dif_clkmgr_result_t;
+
+/**
+ * An Index of a "Gateable" Clock.
+ *
+ * "Gateable" clocks are under full software control: they can be enabled and
+ * disabled by software, which directly starts and stops the identified clock.
+ */
+typedef uint32_t dif_clkmgr_gateable_clock_t;
```

Comment:
I hoped there would be parameters which told me how many clocks were gateable/hintable, but there aren't, so it's not going to be that much fun to validate these indexes.

Reply:
We're not sure what you mean here, can you clarify? The generated enums have a `Last` variant, so isn't this just checking that the value is `< Last` and returning `BadArg` if not?

Reply:
In other DIFs, we normally check that the value of an identifier like this is in the right range (which is normally less than uint32_t) - the common way of doing this in guard clauses is to check against definitions provided by the reggen'd file, for instance `dif_plic_target_t` is always checked to be less than `RV_PLIC_PARAM_NUMTARGET` in `dif_plic.h` - the problem here is there are no parameters from the clockmanager's hjson file that we can check this index against.

We don't use the topgen `Last` definition, because that mixes dif-code with top-specific code, something we aim not to do.

Reply:
We probably could pass these values in the `init` function, I think maybe timer and alert manager does something similar?

Reply:
I've filed #3932 so it's done the same way as the PLIC is, for the moment.